resources/packs/gcp/pubsub.go

```diff
@@ -0,0 +1,347 @@
+
+func (g *mqlGcpProject) GetPubsub() (interface{}, error) {
+	projectId, err := g.Id()
+	if err != nil {
+		return nil, err
+	}
+
+	return g.MotorRuntime.CreateResource("gcp.project.pubsub",
+		"projectId", projectId,
+	)
+}
+
+func (g *mqlGcpProjectPubsubTopic) id() (string, error) {
+	return g.Id()
+}
+
+func (g *mqlGcpProjectPubsubTopicConfig) id() (string, error) {
+	return g.Id()
+}
+
+func (g *mqlGcpProjectPubsubTopicConfigMessagestoragepolicy) id() (string, error) {
+	return g.Id()
+}
+
+func (g *mqlGcpProjectPubsubSubscription) id() (string, error) {
+	return g.Id()
+}
+
+func (g *mqlGcpProjectPubsubSubscriptionConfig) id() (string, error) {
+	return g.Id()
+}
+
+func (g *mqlGcpProjectPubsubSubscriptionConfigPushconfig) id() (string, error) {
+	return g.Id()
+}
+
+func (g *mqlGcpProjectPubsubSnapshot) id() (string, error) {
+	return g.Id()
+}
+
+func (g *mqlGcpProjectPubsub) GetTopics() ([]interface{}, error) {
+	projectId, err := g.ProjectId()
+	if err != nil {
+		return nil, err
+	}
+
+	provider, err := gcpProvider(g.MotorRuntime.Motor.Provider)
+	if err != nil {
+		return nil, err
+	}
+
+	creds, err := provider.Credentials(pubsub.ScopePubSub)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx := context.Background()
+
+	pubsubSvc, err := pubsub.NewClient(ctx, projectId, option.WithCredentials(creds))
+	if err != nil {
+		return nil, err
+	}
+	defer pubsubSvc.Close()
+
+	var topics []interface{}
+
+	it := pubsubSvc.Topics(ctx)
+	for {
+		t, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		mqlTopic, err := g.MotorRuntime.CreateResource("gcp.project.pubsub.topic",
+			"id", t.ID(),
+			"projectId", projectId,
+			"name", t.ID(),
+		)
+		if err != nil {
+			return nil, err
+		}
+		topics = append(topics, mqlTopic)
+	}
+
+	return topics, nil
+}
+
+func (g *mqlGcpProjectPubsubTopic) GetConfig() (interface{}, error) {
+	name, err := g.Name()
+	if err != nil {
+		return nil, err
+	}
+
+	projectId, err := g.ProjectId()
+	if err != nil {
+		return nil, err
+	}
+
+	provider, err := gcpProvider(g.MotorRuntime.Motor.Provider)
+	if err != nil {
+		return nil, err
+	}
+
+	creds, err := provider.Credentials(pubsub.ScopePubSub)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx := context.Background()
+
+	pubsubSvc, err := pubsub.NewClient(ctx, projectId, option.WithCredentials(creds))
+	if err != nil {
+		return nil, err
+	}
+	defer pubsubSvc.Close()
+
+	t := pubsubSvc.Topic(name)
+	cfg, err := t.Config(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	messageStoragePolicy, err := g.MotorRuntime.CreateResource("gcp.project.pubsub.topic.config.messagestoragepolicy",
+		"id", fmt.Sprintf("%s/config/messagestoragepolicy", t.ID()),
+		"allowedPersistenceRegions", core.StrSliceToInterface(cfg.MessageStoragePolicy.AllowedPersistenceRegions),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return g.MotorRuntime.CreateResource("gcp.project.pubsub.topic.config",
+		"id", fmt.Sprintf("%s/config", t.ID()),
```

Comment:
why not `topicID`? The id you use here should be implemented in line 51

```
func (g *mqlGcpProjectPubsubTopicConfig) id() (string, error) {
    topicID, err := g.TopicId()
    if err != nil { return "", err }
    return "gcp.project.pubsub.topic.config/" + topicID + /config
}
```

The implementation above assumes that there is only one config per topic

Reply:
it is, but that breaks if there are topics with the same name in multiple projects. To make this fully unique, I need to add the project as well